File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left = 0
        right = 1
        maxP = 0
        
        while right < len(prices):
            if prices[left] < prices[right]:
                profit = prices[right] - prices[left]
                maxP = max(maxP, profit)
            else:
                left = right
            right += 1
        
        return maxP
    
# Time Complexity: O(n)
# Space Complexity: O(1)    
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int minShare = Integer.MAX_VALUE;

        for(int price : prices){
            minShare = Math.min(price, minShare);

            if(price > minShare){
                maxProfit = Math.max(maxProfit, price - minShare);
            }
        }
        
        return maxProfit;
    }
}

// Time Complexity :O(n)
// Space Complexity: O(1)